Enhancement format command

orhovy opened the pull request:
## Status
~~In Hold ,Reason:Waiting for incidenttypes validator: https://github.com/demisto/demisto-sdk/pull/242/~~
WIP

## Related Issues
fixes: https://github.com/demisto/etc/issues/22489
fixes: https://github.com/demisto/etc/issues/22454
fixes: https://github.com/demisto/etc/issues/21573
~~also fixes: https://github.com/demisto/etc/issues/22816~~ @bakatzir 
## Description
1. Add new format options for :
- incidentfields
- incidenttypes
- indicator fields
- indicator types
- layouts
- dashboards
for each file , remove fields that are not set in schema of this type of file.
2. fixed tags of --source_file -> --input and --output_file -> --output
3. Using --input , if file, format will be activated on it, if directory (Pack or other) , will go through all files in directory and format all yml, json files , Using a generic function.
4. Removed flag --type-file , format command will decide on the type using generic function.
5. New --from-version flag, will set fromversion field to givven , if None,
 On modified files will set to 1.0.0 , On Added files will set to 5.0.0

guyfreund commented:
@orhovy my current work depends on this PR. Do you have an ETA for merging?

bakatzir commented:
We will not release SDK until Thursday(unless a crisis), if you think yana you won't make it , i'll do it in a separate PR, up to you.

bakatzir commented:
@orhovy FYI I merged the incidenttypes validator: #242

anara123 commented on demisto_sdk/__main__.py:
```diff
@@ -323,23 +323,27 @@ def lint(config, dir, **kwargs):
+# -o --out
 @click.option(
-    "-o", "--output-file-name", help="The path where the formatted file will be saved to")
+    "-p", "--pack", help="Specify directory of the pack")
```
I think we agreed that we will have only `in` and `out` 
The code will understand if it is a pack or file
@bakatzir @orhovy

bakatzir replied:
we said input/output(in is a saved word)

anara123 commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,126 @@
+        self.old_file = old_file
+        if not self.source_file:
+            print_color('Please provide <source path>, <optional - destination path>.', LOG_COLORS.RED)
+            sys.exit(1)
```
sys.exit(1) is a bad error handling. raise an exception instead

anara123 commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,126 @@
+        Attributes:
+            input (str): the path to the file we are updating at the moment.
+            output (str): the desired file name to save the updated version of the YML to.
+            yml_data (Dict): YML file data arranged in a Dict.
```
this class doesnt include these attributes

bakatzir commented:
relates to: https://github.com/demisto/etc/issues/22757

idovandijk commented:
@orhovy just a reminder to please tell me before you merge, so I can test it on my environment

bakatzir commented:
also fixes: https://github.com/demisto/etc/issues/22349

idovandijk commented:
I tested this. I know Yana didn't work on playbooks in this PR but I used this opportunity to test them too as this is our main work.

In addition to the files failing validation after `Format`, we also have the issue that Yana mentioned above:
Playbooks are exported without fromversion. We copy+paste that into content. Format will add 1.0.0 which **will make us overlook cases** like changing `fromversion` from 5.0 or 4.1 to 1.0.0.
**The same goes for incidentfields and the rest of the file types.**

Additionally, if we are moving to using packs, `Format` should not create a new file with `playbook-` prefix when used on a file outside of content (currently it does).

As for my tests:
* Results
7/8 playbooks passed
2/2 incident fields passed
0/1 incident types passed
0/2 incident layouts passed

I am attaching the ones that failed, with screenshots:

* Playbooks
[Calculate_Severity_-_Critical_Assets_v2_dev_bugfix.zip](https://github.com/demisto/demisto-sdk/files/4350612/Calculate_Severity_-_Critical_Assets_v2_dev_bugfix.zip)
![Calculate_Severity_-_Critical_Assets_v2_dev_bugfix](https://user-images.githubusercontent.com/43602124/76993116-05489180-6955-11ea-9897-15c8d4ae57bd.PNG)

* Incident Types
[incidenttype-Cortex_XDR_Port_Scan.zip](https://github.com/demisto/demisto-sdk/files/4350615/incidenttype-Cortex_XDR_Port_Scan.zip)
![incidenttype-Cortex_XDR_Port_Scan](https://user-images.githubusercontent.com/43602124/76993197-227d6000-6955-11ea-86ff-200aa8472c9f.PNG)


* Layouts:
[layout-details-Employee_Offboarding-V2.zip](https://github.com/demisto/demisto-sdk/files/4350628/layout-details-Employee_Offboarding-V2.zip)
![layout-details-Employee_Offboarding-V2](https://user-images.githubusercontent.com/43602124/76993245-37f28a00-6955-11ea-8cd4-d1bd6d429ca5.PNG)
[layout-details-Phishing-V2.zip](https://github.com/demisto/demisto-sdk/files/4350621/layout-details-Phishing-V2.zip)
![layout-details-Phishing-V2](https://user-images.githubusercontent.com/43602124/76993303-4d67b400-6955-11ea-9696-d4a9a7666f90.PNG)


@michalgold FYI

orhovy commented:
@idovandijk  fixed the layout, test should now pass.
about incidentType test ,  I tried to reproduce the failure a couple of times and I saw that this kind of error is written only when The name of the file is different from the id/name inside the file. It seems like by accident you pasted a Incident to Cortex XDR Port Scan inside the file  Packs/CortexXDR/IncidentTypes/Cortex_XDR_Incident.json. Can you please try again this one test?

bakatzir commented on demisto_sdk/commands/format/tests/formatting_js_test.py:
```diff
@@ -0,0 +1,33 @@
+from demisto_sdk.commands.format.format_module import format_manager
+
+
+class TestFormattingJson:
```
if not a good reason for the class, you can rm.

bakatzir replied:
also, let's rename the file to `formatting_json_test` as we refer internally usually js to javascript.

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+
+    DEFAULT_JSON_VERSION = -1
+    NEW_FILE_DEFAULT_FROMVERSION = '5.0.0'
+    OLD_FILE_DEFAULT_FROMVERSION = '1.0.0'
```
you have defined it in a different file.
consider putting all constants in the same file, e.g in `constants.py`

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+    NEW_FILE_DEFAULT_FROMVERSION = '5.0.0'
+    OLD_FILE_DEFAULT_FROMVERSION = '1.0.0'
+
+    def __init__(self, input='', output='', old_file='', path='', from_version=''):
```
use type hints in all functions

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+
+        return os.path.join(source_dir, file_name)
+
+    def get_json_data_as_dict(self):
```
I believe we have this function(or a variance of it) in `tools.py`

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+        self.output_file_name = self.set_output_file_name(output)
+        self.arguments_to_remove = self.arguments_to_remove()
+
+    def set_output_file_name(self, output_file_name):
```
what is the logic to have one here and one in the generic_yml?

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+        source_dir = os.path.dirname(self.source_file)
+        file_name = os.path.basename(output_file_name or self.source_file)
+
+        if self.__class__.__name__ == 'PlaybookYMLFormat':
```
how can it be, isn't a playbook is of type `yml`?

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+        with open(self.source_file) as file:
+            return json.load(file)
+
+    def set_version_to_default(self):
```
we have it also in `update_generic_yml`. `set_version_to_default`
let's think together how we put all common functions in between this classes in one shared location.

orhovy replied:
Moved this function to update_generic, however  update_generic but update_generic_yml runs it over as if in each yml id and name are under different keys.

bakatzir commented on demisto_sdk/commands/format/update_generic_json.py:
```diff
@@ -0,0 +1,152 @@
+            print_color('The files are invalid', LOG_COLORS.RED)
+            return 1
+
+    def format_file(self):
```
what is the meaning of this?

bakatzir commented on demisto_sdk/commands/common/constants.py:
```diff
@@ -932,7 +933,7 @@ class PB_Status:
-    'incidenttype': [INCIDENT_TYPE_REGEX],
+    'incidenttype': JSON_ALL_INCIDENT_TYPES_REGEXES,
     'image': [IMAGE_REGEX],
     'reputation': [REPUTATION_REGEX],
```
the reputations have moved to here: `https://github.com/demisto/content/tree/master/Packs/Base/Misc`
please validate that we are using the correct regexes.

orhovy replied:
Changed it in constants of MiscDir

bakatzir commented on demisto_sdk/commands/format/update_generic_yml.py:
```diff
@@ -1,111 +1,65 @@
+        if self.data.get('display'):
+            self.data['display'] = self.data.get('display', '').replace('_copy', '').replace('_dev', '')
 
     def update_id_to_equal_name(self):
```
also, maybe we can change to `update_id_to_equal_field` and then move the field name. from the talk with Anar for incidentfield we can move `cliname` - just a thought. up to you.

bakatzir commented on demisto_sdk/commands/format/tests/formatting_jsond_test.py:
```diff
@@ -0,0 +1,33 @@
+import os
```
typo in filename

bakatzir commented on demisto_sdk/commands/format/tests/formatting_test.py:
```diff
@@ -10,43 +10,52 @@
 from demisto_sdk.commands.format.update_integration import IntegrationYMLFormat
```
change filename to formatting_yml_test?

bakatzir commented on demisto_sdk/commands/format/update_incidentfields.py:
```diff
@@ -0,0 +1,33 @@
+from demisto_sdk.commands.format.update_generic_json import BaseUpdateJSON
+from demisto_sdk.commands.common.hook_validations.incident_field import IncidentFieldValidator
+
+ARGUMENTS_DEFAULT_VALUES = {
```
lets think of a way to have this at only one place. since if a field will have to be added, we do not want to add it in all the `update_<json_entity>.py` files.

bakatzir commented on demisto_sdk/commands/format/update_layout.py:
```diff
@@ -0,0 +1,59 @@
+            output (str): the desired file name to save the updated version of the YML to.
+    """
+
+    def __init__(self, input='', output='', path='', from_version=''):
```
do not forget type annotation

bakatzir commented on demisto_sdk/commands/format/update_playbook.py:
```diff
@@ -9,22 +9,20 @@ class PlaybookYMLFormat(BaseUpdateYML):
 
-    def __init__(self, source_file='', output_file_name=''):
-        super().__init__(source_file, output_file_name)
+    def __init__(self, input='', output='', path='', from_version=''):
```
add type annotation

orhovy commented on demisto_sdk/commands/format/format_module.py:
```diff
@@ -1,49 +1,82 @@
+    error_list = []
+    schema_path = os.path.normpath(
+        os.path.join(__file__, "..", "..", "common", SCHEMAS_PATH, '{}.yml'.format(file_type)))
+    res = FILE_TYPE_AND_LINKED_CLASS[file_type](input=input, path=schema_path,
```
@bakatzir how should I make it more clearer, a comment or change name of dict?

orhovy commented on demisto_sdk/commands/format/update_generic_yml.py:
```diff
@@ -1,111 +1,65 @@
-        return self.yml_data.get(path, self.yml_data)
+        return self.data.get(path, self.data)
 
     def remove_copy_and_dev_suffixes_from_name(self):
```
moved it

orhovy commented on demisto_sdk/commands/format/format_module.py:
```diff
@@ -1,49 +1,82 @@
+    res = FILE_TYPE_AND_LINKED_CLASS[file_type](input=input, path=schema_path,
+                                                from_version=from_version, **kwargs).format_file()
+    if res:
+        error_list.append(f'Failed to format {input}.' + file_type)
```
Added log with all successful, failures and skipped files

orhovy commented on demisto_sdk/commands/format/update_classifier.py:
```diff
@@ -0,0 +1,17 @@
+from demisto_sdk.commands.format.update_generic_json import BaseUpdateJSON
+
+
+class ClassifierJSONFormat(BaseUpdateJSON):
```
Added

orhovy commented on demisto_sdk/commands/format/update_generic.py:
```diff
@@ -0,0 +1,144 @@
+            source_file (str): the path to the file we are updating at the moment.
+            output_file (str): the desired file name to save the updated version of the YML to.
+            relative_content_path (str): Relative content path of output path.
+            old_file (bool): Whether the file is a added file = new or a modified file = old.
```
you are right, I changed it.

orhovy commented on demisto_sdk/commands/format/update_generic_yml.py:
```diff
@@ -1,111 +1,65 @@
-        self.id_and_version_location['id'] = self.yml_data['name']
+        self.id_and_version_location['id'] = self.data['name']
 
     def set_version_to_default(self):
```
Moved this function to update_generic, however  update_generic but update_generic_yml runs it over as if in each yml id and name are under different keys.

orhovy commented on demisto_sdk/commands/format/update_layout.py:
```diff
@@ -0,0 +1,83 @@
+    def __init__(self, input: str = '', output: str = '', path: str = '', from_version: str = '', no_validate: bool = False):
+        super().__init__(input, output, path, from_version, no_validate)
+
+    def set_version_to_default(self):
```
layout needs a different function as version is under layout -> version

bakatzir replied:
Ok, let's add a comment regarding it.

orhovy commented:
@anara123  @bakatzir 
I ran a few tests from server on json files.
Downloaded using custom content all kinds of files:
incidentfield , incidenttype , indicatortype , layout ,dashboard
I attached here the test files, in each folder of file type there is the file downloaded from server with the ending _fromServer and the formatted file ending with _format.
[format_tests.zip](https://github.com/demisto/demisto-sdk/files/4388822/format_tests.zip)

anara123 commented:
@orhovy looks good , lets merge

idovandijk commented:
I'll test it as well today with the files that previously failed

idovandijk commented:
Ran format on a local file on my system. It skipped validation on it
![image](https://user-images.githubusercontent.com/43602124/77844646-04afc680-71b1-11ea-89cd-2b3f109a357d.png)

The expected (and previous) behavior is that it works outside of the repo folder too.
When I moved the file to the repo folder and re-ran, I still got an error:

![image](https://user-images.githubusercontent.com/43602124/77844749-c6ff6d80-71b1-11ea-8cad-6fc1263b865a.png)

I don't know why validate says failed, I see no error, but I suspect it's the backslash, @yaakovi is this backslash supposed to be OK?
The failing file is this:
[Calculate_Severity_-_Critical_Assets_v2_dev_bugfix.zip](https://github.com/demisto/demisto-sdk/files/4398475/Calculate_Severity_-_Critical_Assets_v2_dev_bugfix.zip)

bakatzir commented:
@idovandijk 
1. Validate can't run outside of content repo - hence it is skipped. that's the expected behaviour, but we can furhter explain this message.
2. if fails in the content repo because the format changes the file ID to be the file name, which is a break of BC for playbooks - we need to look at this @orhovy.
3. the printing with the backslash does look weird, but it runs the validation on the file so it's ok.

idovandijk commented:
> @idovandijk
> 
> 1. format can run anywhere, Validate can't run outside of content repo - hence it is skipped. that's the expected behaviour, but we can furhter explain this message.
> 2. if fails in the content repo because the format changes the file ID to be the file name, which is a break of BC for playbooks - we need to look at this @orhovy.
> 3. the printing with the backslash does look weird, but it runs the validation on the file so it's ok.

OK. I will test on the rest of the files after the change

idovandijk commented:
Apparently previous error was due to having copied the newly exported file (with a different ID than the one in master) to the repo. I am re-testing now

idovandijk commented:
Almost everything worked. There are 2 issues. One is with a new playbook (that we haven't tested before):
[Failed_Login_Playbook_-_Slack_v2_Yarden_Changes.zip](https://github.com/demisto/demisto-sdk/files/4398518/Failed_Login_Playbook_-_Slack_v2_Yarden_Changes.zip)

![Slack v2 1](https://user-images.githubusercontent.com/43602124/77845327-b7365800-71b6-11ea-874a-c8335fd79308.png)

The other one is that we are not adding, or asking to add `tests: ` section or at least keeping the one that existed before in the file.

Other than that, all the files that previously did not work - now work :)

orhovy commented:
tested it on all files, added better description in code about from version addition , will add in readme file as well.
Talked to @idovandijk , added a copy of tests key in yml files if exist in old file.
-About fromVersion in the playbook, as I see, such file name does not exist in content repo , as a result the format command assumes that it is a new playbook and adds fromversion 5.0.0 to such files.

yaakovi commented on demisto_sdk/commands/common/constants.py:
```diff
@@ -255,7 +255,7 @@ def found_hidden_param(parameter_name):
 LAYOUTS_DIR = 'Layouts'
 CLASSIFIERS_DIR = 'Classifiers'
-MISC_DIR = 'Misc'
+MISC_DIR = 'Packs/Base/Misc'
```
@orhovy @bakatzir why was this changed?
it broke the `init` command

orhovy replied:
It was changed because the directory of reputation files was changed to this path. In order the validate will validate reputation files(catch their regex), this was changed.

bakatzir replied:
misc dir was re changed in content and is not in base pack anymore: https://github.com/demisto/demisto-sdk/pull/322

yaakovi replied:
You are right but this is not relevant to the `init` command.
Even before the revert this is breaking the command

anara123 replied:
I filed an issue https://github.com/demisto/etc/issues/23293
